### ISSUE TITLE
Scaling langevin noise

### ISSUE THREAD

**Issue by maltet96** [JavierAntoran/Bayesian-Neural-Networks#17]
Hi @JavierAntoran  @stratisMarkou ,

I am currently using SGLD in another project. I have some example code, where the noise added to the parameters is scaled by the learning rate:

`noise = torch.randn(n.size()) * param_noise_sigma * lr`.

I belive, this is wrong and so I scaled the noise with the square root of the learning rate:

`noise = torch.randn(n.size()) * np.sqrt(lr)`

This works well. However, you scale the learning rate by 

`1/np.sqrt(sigma)`

If I do this, the noise gets way too big and the network does not learn. Also, in the initial paper by Welling et al. it seems to me that the first approach is right, though I already found out that the variance of the noise scales with lr**2, which is why the square root is necessary.

I hope you can explain to me, why the ()^-1 is necessary. Thank you in advance!

**Comment by JavierAntoran** [comment]
Hi @maltet96,

By default pytorch uses the `Tensor.add_` method to update weights.  The first argument is multiplied by the second argument and added to the tensor. In this case -group['lr'] is going to multiply both the weights' gradients and the additive noise. As you correctly point out, we need to multiply unit variance noise by the square root of the learning rate such that the resulting variance matches the learning rate. Thus, in the previous step, we divide the noise term by the square root of the learning rate.

    langevin_noise = p.data.new(p.data.size()).normal_(mean=0, std=1) / np.sqrt(group['lr'])
    p.data.add_(-group['lr'], 0.5 * d_p + langevin_noise)

**Comment by maltet96** [comment]
I do understand, thank you very much for explaining it!
